framework/decode/vulkan_virtual_swapchain.cpp

```diff
@@ -458,56 +689,80 @@ VkResult VulkanVirtualSwapchain::QueuePresentKHR(PFN_vkQueuePresentKHR
+    // TODO: There is a potential issue here where a vkQueuePresent comes in on a queue (let's call
+    // it QueueX) which does not support vkCmdCopyImage (i.e. a video-only queue).  In that case,
+    // we would need to insert an emtpy command buffer into the command stream of QueueX which
+    // triggers a semaphore (let's say SemA), then we would need to submit the vkCmdCopyImage in a
+    // command buffer on a queue that supports it (let's say QueueY) which will wait on SemA to
+    // start and signaling another semaphore (SemB) when it is done.  Then, we need to add the
+    // QueuePresent to QueueX, but waiting on SemB before it executes.  And that is assuming that
+    // the buffer image is even accessible on both Queues!
+
+    for (uint32_t i = 0; i < swapchainCount; ++i)
+    {
+        const auto* swapchain_info      = swapchain_infos[i];
+        uint32_t    capture_image_index = capture_image_indices[i];
+        uint32_t    replay_image_index  = present_info->pImageIndices[i];
 
-    VkOffset3D offsets[2] = {
+        // Get the private data so we have access to the virtual swapchain-specific information.
+        if (private_data_.find(swapchain_info->handle) == private_data_.end())
         {
-            0,
-            0,
-            0,
-        },
-        {
-            0,
-            0,
-            1,
-        },
-    };
+            GFXRECON_LOG_ERROR("Virtual swapchain vkQueuePresentKHR missing private data for swapchain (ID = %" PRIu64
+                               ")",
+                               swapchain_info->capture_id);
+        }
+        PrivateData* private_data = private_data_[swapchain_info->handle];
+        assert(private_data != nullptr);
 
-    auto                     length = present_info->swapchainCount;
-    std::vector<VkSemaphore> semaphores;
+        const auto& virtual_image = private_data->virtual_swapchain_images[capture_image_index];
+        const auto& replay_image  = private_data->replay_swapchain_images[replay_image_index];
 
-    for (uint32_t i = 0; i < length; ++i)
-    {
-        uint32_t    capture_image_index = capture_image_indices[i];
-        uint32_t    replay_image_index  = present_info->pImageIndices[i];
-        const auto* swapchain_info      = swapchain_infos[i];
-        const auto& virtual_image       = swapchain_info->virtual_images[capture_image_index];
-        const auto& swapchain_image     = swapchain_info->swapchain_images[replay_image_index];
+        // Use a command buffer and semaphore from the same queue index
+        auto command_buffer = private_data->copy_command_buffers[queue_family_index][capture_image_index];
+        auto copy_semaphore = private_data->copy_semaphores[queue_family_index][capture_image_index];
+
+        std::vector<VkSemaphore> wait_semaphores;
+        std::vector<VkSemaphore> signal_semaphores;
+
+        // Only wait on the first copy
+        if (i == 0)
+        {
+            if (present_info->waitSemaphoreCount > 0)
```

Comment:
You can replace all this with a call to `std::fill` or `std::fill_n` with `back_inserter` (https://en.cppreference.com/w/cpp/iterator/back_inserter)

Reply:
That does not work for C style arrays.  Instead I can do the following since it's a first assignment:

```
wait_semaphores.assign(present_info->pWaitSemaphores,
                       present_info->pWaitSemaphores + present_info->waitSemaphoreCount);
```

Reply:
std::fill does work on C arrays but I agree if you have something that works well and is idiomatic, there's no need to change it.

Reply:
I see - I think the original code I commented on was filling a `vector<bool>` but it's changed since then to be an assignment into a vector and for that case I think `assign` is great!

Reply:
Done